draft-ietf-rats-corim.md

```diff
@@ -2070,133 +2042,195 @@ Logically, new ECT entries are appended to the existing ACS.
 But implementations may optimize ECT order to achieve better performance.
 Additions to the ACS MUST be atomic.
 
-## ECT Comparison {#sec-ect-compare}
+## Comparing a condition ECT against the ACS {#sec-match-condition-ect}
 
 [^issue] https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71
 
-This specification defines the comparison algorithm for the codepoints described in sub-sections below.
-A CoRIM profile may define additional negative codepoints and their matching algorithms.
-Specifications that extend CoMID MUST also define comparison algorithms for their added codepoints.
-If a codepoint's comparison algorithm is not stated or does not default to the comparison algorithms described in this specification, then the Verifier MUST compare the binary equality of the CBOR encodings of the values.
+The algorithm used to compare a condition ECT against the ACS is stateless; it depends only on the condition ECT being compared and the contents of the ACS at the time of the comparison.
+
+A Verifier SHALL iterate over all ACS entries and SHALL attempte to match the condition ECT against each ACS entry using the algorithm described in {{sec-match-one-condition-ect}}.
+A Verifier SHALL create a "matched entries" set, and SHALL populate it with all ACS entries which matched the condition ECT.
+
+If the matched entries array is not empty, then the condition ECT matches the ACS.
+
+If the matched entries array is empty, then the condition ECT does not match the ACS.
+
+### Comparing a condition ECT against a single ACS entry {#sec-match-one-condition-ect}
+
+A Verifier SHALL perform all of the comparisons in sections {{sec-compare-environment}), {{sec-compare-authority}), {{sec-compare-cm}} and {{sec-compare-element-list}).
```

Comment:
I am not expecting evidence to always be marked with a profile, so doing a profile comparison first will sometimes fail.
If we have a profile comparison then should it be the other way round - a candidate ECT without a profile will match a condition ECT with a profile?

Reply:
Hmmmmmmmm that is very tricky indeed. That would mean that we have to be very careful that any CBOR that matches the base CDDL has a solid semantics that cannot be extended by a profile, such as with additional value encodings that allow for weaker pattern matching (e.g., adding any-of to standard codepoints if it weren't already there).

Say we have a condition ECT that has conditional endorsement series triple that on condition 3 includes nonstandard values, and condition 4 has standard values again and is the one we expect to match.
Do we permit the condition with a profile to be evaluated against the candidate without a profile until we get to the conditional series additions?

I understand -1 would never compare true to the nonnegative standard codepoints, but if there are extensible type choices in standard codepoints, we have to ensure that they are compared with encoding equality only to avoid overriding the base CoRIM semantics. `$version-scheme` for example permits any `int`, so how do we allow a profile to define a new one without messing with the base CoRIM semantics? I don't know. We haven't defined the semantics of version matching.

Reply:
My expectation is if a profile defines an alternative matching algorithm from one defined in the "base" corim spec, the profile algorithm takes precedence.  A simple example might be svn comparisons (not min-svn) where current algorithm is equivalence, a RV profile might apply range semantics such that an actual svn=2 might be accepted by an algorithm that said (svn >= 1). Nevertheless, the ACS still contains svn=2 and any condition endorsement statement that compares for equality still uses the base spec's equality check algorithm. It suggests a constraint on profiles that the syntax for describing alternative matching should have a well-defined format. For example, the Intel profile defines a CBOR tag that prefaces the matching condition statement. Hence, the Verifier is not undecided about which algorithm to apply even if processing order varies, so long as the profile is parsed before processing begins.

Reply:
svn-type-choice is not extensible, so your example doesn't track for me.

We shouldn't ever have certain condition tags in the ACS as actual state, such as `tagged-min-svn`.
Are we going to allow profiles to extend measurement-values-map values with their own pattern languages for more complex comparisons against the ACS? It becomes a kind of stored-procedure input at that point. It could be useful, but we can't permit for conditions to leak into actual state.

Reply:
I think the profile should take precedence if it exists, but that doesn't mean that we need to compare the profile between the condition ECT and the ACS entry.

Can we say something like this:
```suggestion
If the condition ECT contains a profile then the documentation for that profile describes the matching algorithm.
Profile authors SHOULD make their matching algorithms similar to the algorithm described in this document where possible.

A Verifier SHALL perform all of the comparisons in sections {{sec-compare-environment}), {{sec-compare-authority}), {{sec-compare-cm}} and {{sec-compare-element-list}).
```

This means that the profile can add new matching algorithms, like the ANY_OF<> algorithm proposed in another PR.